docs/rfc/rfc-004-accounts.md

```diff
@@ -0,0 +1,323 @@
+- Provide a way to query information that cannot be processed from raw state reflection, ex: compute information from lazy
+  state that has not been yet concretely processed (eg: balances with respect to lazy inputs/outputs)
+
+#### Schema
+
+Schema provides the definition of an account from `API` perspective, and it's the only thing that should be taken into account
+when interacting with an account from another account or module, for example: an account is an `authz-interface` account if
+it has the following message in its execution messages `MsgProxyStateTransition{ state_transition: google.Protobuf.Any }`.
+
+### Migrate
+
+Migrate defines the entrypoint that allows an `Account` to migrate its state from a previous version to a new one. Migrations
+can be initiated only by the account itself, concretely this means that the migrate action sender can only be the account address
+itself, if the account wants to allow another address to migrate it on its behalf then it could create an execution message
+that makes the account migrate itself.
+
+### x/accounts module
+
+In order to create accounts we define a new module `x/accounts`, note that `x/accounts` deploys account with no authentication
+credentials attached to it which means no action of an account can be incepted from a TX, we will later explore how the
+`x/authn` module uses `x/accounts` to deploy authenticated accounts.
+
+This also has another important implication for which account addresses are now fully decoupled from the authentication mechanism
+which makes in turn off-chain operations a little more complex, as the chain becomes the real link between account identifier
+and credentials.
+
+We could also introduce a way to deterministically compute the account address.
+
+Note, from the transaction point of view, the `init_message` and `execute_message` are opaque `google.Protobuf.Any`.
+
+The module protobuf definition for `x/accounts` are the following:
+
+```protobuf
+// Msg defines the Msg service.
+service Msg {
+  rpc Deploy(MsgDeploy) returns (MsgDeployResponse);
+  rpc Execute(MsgExecute) returns (MsgExecuteResponse);
+  rpc Migrate(MsgMigrate) returns (MsgMigrateResponse);
+}
+
+message MsgDeploy {
+  string sender = 1;
+  string kind = 2;
+  google.Protobuf.Any init_message = 3;
+  repeated google.Protobuf.Any authorize_messages = 4 [(gogoproto.nullable) = false];
+}
+
+message MsgDeployResponse {
+  string address = 1;
+  uint64 id = 2;
+  google.Protobuf.Any data = 3;
+}
+
+message MsgExecute {
+  string sender = 1;
+  string address = 2;
+  google.Protobuf.Any message = 3;
+  repeated google.Protobuf.Any authorize_messages = 4 [(gogoproto.nullable) = false];
+}
+
+message MsgExecuteResponse {
+  google.Protobuf.Any data = 1;
+}
+
+message MsgMigrate {
+  string sender = 1;
+  string new_account_kind = 2;
+  google.Protobuf.Any migrate_message = 3;
+}
+
+message MsgMigrateResponse {
+  google.Protobuf.Any data = 1;
+}
+
+```
+
+#### MsgDeploy
```

Comment:
I'd like MsgDeploy to be optional. I can see how this is needed for stateful accounts, but many account implementations can be stateless (i.e.: priv/pubkeys). I think there is value in allowing users to interact with accounts that haven't been "deployed". 

If the only objective of an account deploy is to make it easy to check its existence (someone has interacted with it), then we can ensure deploy gets called on first interaction if the account doesn't exist.

Reply:
@nicolaslara `MsgDeploy` is needed to provide the initial state for the account, in the same way as `MsgInstantiate` works in cosmwasm.

Reply:
I guess what I want here is to default state to be assumed for an account if it hasn't been deployed. I think there's value in interacting with inexistent accounts the way it can be done now. This gets more complex with account rekeying though

Reply:
the auth module will still exist, i think we are leaning towards x/accounts has explicit avvount creation while x/auth has implicit.

Reply:
So if you want to do an IBC transfer but use the new account type's features you'd have to use x/auth and then migrate right? 

For composability (on ibc-hooks, for example), one thing that we can do is to deploy an x/accounts account as part of what gets executed OnRecv. What we're currently doing on ibc-hooks is sending the tokens to `sha(prefix|channel|counterparyAccount)` and that can only be controlled via ibc-hooks. That could be modified to deploy an account with credentials such that it can be controlled via ibc hooks on that channel or the pubkey of the original account.

Reply:
> So if you want to do an IBC transfer but use the new account type's features you'd have to use x/auth and then migrate right?

if the account doesnt exist most likely.

interesting solution, i didnt think of this. That is pretty good, im not sure who would implement this part though

Reply:
resolving this conversation as we discussed how app devs could have hooks to create account on the fly for a little extra gas